From dfd27892a34130d62b1f96851288be1b195342d2 Mon Sep 17 00:00:00 2001
From: John Chadwick <jchadwick@buf.build>
Date: Wed, 5 Mar 2025 20:34:56 -0500
Subject: [PATCH 2/3] Add missing #include for absl::StrCat

---
 internal/testing.cc | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/internal/testing.cc b/internal/testing.cc
index 77e4c65b..e0dfa75e 100644
--- a/internal/testing.cc
+++ b/internal/testing.cc
@@ -14,6 +14,8 @@
 
 #include "internal/testing.h"
 
+#include "absl/strings/str_cat.h"
+
 namespace cel::internal {
 
 void AddFatalFailure(const char* file, int line, absl::string_view expression,
-- 
2.47.2

